omniscidb/QueryEngine/ArithmeticIR.cpp

```diff
@@ -385,29 +385,61 @@ llvm::Value* CodeGenerator::codegenMul(const hdk::ir::BinOper* bin_oper,
-            adjusted_lhs_lv,
-            cgen_state_->ir_builder_.CreateSDiv(chosen_max, positive_rhs_lv)),
-        // underflow
-        cgen_state_->ir_builder_.CreateICmpSLT(
-            adjusted_lhs_lv,
-            cgen_state_->ir_builder_.CreateSDiv(chosen_min, positive_rhs_lv)));
-    cgen_state_->ir_builder_.CreateCondBr(detected, mul_fail, mul_ok);
+    auto const_one = llvm::ConstantInt::get(rhs_lv->getType(), 1, true);
+
+    // If any of the args was 0, no overflow occurs
+    auto lhs_is_zero = cgen_state_->ir_builder_.CreateICmpEQ(lhs_lv, const_zero);
+    auto rhs_is_zero = cgen_state_->ir_builder_.CreateICmpEQ(rhs_lv, const_zero);
+    auto args_zero = cgen_state_->ir_builder_.CreateOr(rhs_is_zero, lhs_is_zero);
+    cgen_state_->ir_builder_.CreateCondBr(args_zero, mul_ok, mul_check_1);
+    cgen_state_->ir_builder_.SetInsertPoint(mul_check_1);
+
+    // Check the sign of the args
+    auto lhs_is_neg = cgen_state_->ir_builder_.CreateICmpSLT(lhs_lv, const_zero);
+    auto rhs_is_neg = cgen_state_->ir_builder_.CreateICmpSLT(rhs_lv, const_zero);
+    auto args_is_neg = cgen_state_->ir_builder_.CreateOr(lhs_is_neg, rhs_is_neg);
+
+    auto lhs_is_pos = cgen_state_->ir_builder_.CreateICmpSGT(lhs_lv, const_zero);
+    auto rhs_is_pos = cgen_state_->ir_builder_.CreateICmpSGT(rhs_lv, const_zero);
+    auto args_is_pos = cgen_state_->ir_builder_.CreateOr(lhs_is_pos, rhs_is_pos);
+
+    // Get the absolute value of the args
+    auto lhs_neg = cgen_state_->ir_builder_.CreateNeg(lhs_lv);
+    auto rhs_neg = cgen_state_->ir_builder_.CreateNeg(rhs_lv);
+    auto lhs_pos = cgen_state_->ir_builder_.CreateSelect(lhs_is_neg, lhs_neg, lhs_lv);
+    auto rhs_pos = cgen_state_->ir_builder_.CreateSelect(rhs_is_neg, rhs_neg, rhs_lv);
+
+    // lhs and rhs are in [1, 2^n], where n is the number of digits.
+    // Check how the max allowed absolute value (2^n for negative, 2^(n-1) for
+    // positive) divided by an argument compares to the other.
+
+    // if(IsNegative) && UX > (static_cast<U>(std::numeric_limits<T>::max()) + U(1)) / UY;
+    auto max_plus_one = cgen_state_->ir_builder_.CreateAdd(chosen_max, const_one);
+    auto div_limit_plus = cgen_state_->ir_builder_.CreateUDiv(max_plus_one, rhs_pos);
+    auto cmp_plus = cgen_state_->ir_builder_.CreateICmpUGT(lhs_pos, div_limit_plus);
+    auto neg_overflow = cgen_state_->ir_builder_.CreateAnd(args_is_neg, cmp_plus);
+    cgen_state_->ir_builder_.CreateCondBr(neg_overflow, mul_fail, mul_check_2);
+    cgen_state_->ir_builder_.SetInsertPoint(mul_check_2);
+
+    // if !(IsNegative) && UX > (static_cast<U>(std::numeric_limits<T>::max())) / UY;
+    auto div_limit = cgen_state_->ir_builder_.CreateUDiv(chosen_max, rhs_pos);
+    auto cmp = cgen_state_->ir_builder_.CreateICmpUGT(lhs_pos, div_limit);
+    auto pos_overflow = cgen_state_->ir_builder_.CreateAnd(args_is_pos, cmp);
+    cgen_state_->ir_builder_.CreateCondBr(pos_overflow, mul_fail, mul_ok);
```

Comment:
I think there's still a problem with the implementation. The `arg_is_pos`/`arg_is_neg` should have been calculated as `xor` not `or` (as they are in the link that you've shared btw).
Consider the following simplified example:
Let's have a bit width of 3:
| bit representation | signed value | unsigned value |
| --- | --- | --- |
| 000 | 0 | 0 |
| 001 | 1 | 1 |
| 010 | 2 | 2 |
| 011 | 3 | 3 |
| 100 | -4 | 4 |
| 101 | -3 | 5 |
| 110 | -2 | 6 |
| 111 | -1 | 7 |

Consider a * b: a = -2, b = -2. This is a positive overflow case. Then:
* `lhs_is_neg = rhs_is_neg = true`, thus `args_is_neg = true`
* `lhs_is_pos = rhs_is_pos = false`, thus `args_is_pos = false`
* In our case signed max is `3`, so `max_plus_one = 4`, and `div_limit_plus = 4/2 = 2`
* `div_lim = 3/2 = 1`
* `cmp_plus = 2 > 2 = false`, so no negative overflow, but
* `cmp = 2 > 1 = true`, `pos_overflow = false and true = false`, which is incorrect.

I also suspect you don't need the pair of positive/negative cases since you already got rid of the zeroes case in the first basic block. Xoring the arguments signedness will determine the case to check.